velox/expression/tests/SpecialFormExprOptimizationTest.cpp

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gtest/gtest.h"
+
+#include "velox/expression/ConstantExpr.h"
+#include "velox/expression/ExprCompiler.h"
+#include "velox/expression/FieldReference.h"
+#include "velox/expression/RegisterSpecialForm.h"
+#include "velox/functions/prestosql/registration/RegistrationFunctions.h"
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+#include "velox/parse/TypeResolver.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::test;
+using namespace facebook::velox::exec;
+using functions::test::FunctionBaseTest;
+
+class SpecialFormExprOptimizationTest : public FunctionBaseTest {
+ protected:
+  static void SetUpTestCase() {
+    parse::registerTypeResolver();
+    exec::registerFunctionCallToSpecialForms();
+    functions::prestosql::registerAllScalarFunctions();
+    memory::MemoryManager::testingSetInstance({});
+  }
+
+  const core::QueryConfig config_{{}};
+};
+
+TEST_F(SpecialFormExprOptimizationTest, conjunct) {
+  // true AND false AND a -> false
+  auto compiledExpr =
+      compileExpression("(1 = 1) AND (2 != 2) AND a", ROW({"a"}, {BOOLEAN()}))
+          ->exprs()
+          .front();
+  ASSERT_TRUE(compiledExpr->isConstant());
+  ASSERT_EQ(compiledExpr->type(), BOOLEAN());
+  assertEqualVectors(
+      vectorMaker_.constantVector<bool>({false}),
+      compiledExpr->as<ConstantExpr>()->value());
+
+  // false OR true OR a -> true
+  compiledExpr =
+      compileExpression("(1 != 1) OR (2 = 2) OR a", ROW({"a"}, {BOOLEAN()}))
+          ->exprs()
+          .front();
+  ASSERT_TRUE(compiledExpr->isConstant());
```

Comment:
The code to verify that compiledExpr is a constant repeats a couple times. Consider refactoring.

```
assertConstantTrue(compiledExpr);
assertConstantFalse(compiledExpr);
...
void assertConstantTrue(expr) {
   assertConstantBoolean(expr, true);
}
```